### PR TITLE
Create wrangle.py

### PR DESCRIPTION
**Pull request by mr-devs**
Addressing #70, this PR creates some data wrangling functions that I thought could be helpful. 

_Note_: I have not been able to actually run the `tests.py` script because I am maxed out on my monthly tweets so it will fail no matter what. Please look that script over carefully or give it a go on your end. Either way, it shouldn't break anything old because it's a separate module.

I was originally intending to create functions like `wrangle.extract(tweet_obj, "hashtags")` - which would take a tweet object and return all the hashtags - but this seems too specific given Twitter's new data format. I ended up going with a handful of general functions that I felt could be applicable/useful for any bit of Twitter data (I think?).

Of course, please share your thoughts on whether these are worth including or not.

### CODE REVIEW

**Comment by mr-devs** [comment]
FYI, I will update the Wiki with examples that utilize actual tweet data after this is merged.

File: osometweet/wrangle.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
+from collections.abc import MutableMapping
+
+def _flatten_dict_gen(d, parent_key, sep):
+    for k, v in d.items():
+        new_key = parent_key + sep + k if parent_key else k
```
Comment (yangkcatiu) [inline osometweet/wrangle.py:5]:
```py
new_key = f"{parent_key}{sep}{k}" if parent_key else k
```
Probably safer this way since `parent_key` and `k` might not be str.

File: osometweet/wrangle.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    ~~~
+    ### 3. Changing `sep`
+    # This string is what will separate key path strings
+    flatten_dict(dictionary, sep = "/")
+    
+    # Returns
+    {'a': 1, 'b/c': 2, 'b/d': 5, 'e/f': 4, 'e/h': 3, 'i': 3}
+
+    """
+    if not isinstance(dictionary,dict):
```
Comment (yangkcatiu) [inline osometweet/wrangle.py:76]:
```py
if not isinstance(dictionary, dict):
```
Just a format thing

File: osometweet/wrangle.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    }
+    
+    # Call get_dict_paths
+    print(list(get_dict_paths(dictionary)))
+    
+    # Returns
+    [['a'], ['b', 'c'], ['b', 'd'], ['e', 'f'], ['e', 'g'], ['h']]
+    
+    """
+    if not isinstance(path,list):
```
Comment (yangkcatiu) [inline osometweet/wrangle.py:121]:
```py
if not isinstance(path, list):
```
Just a format thing

**Review by yangkcatiu — COMMENTED** [review]
BTW, you can use

```sh
python tests.py TestWranlge
```
To just run the test you need

File: osometweet/wrangle.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    {'a': 1, 'b/c': 2, 'b/d': 5, 'e/f': 4, 'e/h': 3, 'i': 3}
+
+    """
+    if not isinstance(dictionary, dict):
+        raise TypeError("`dictionary` must be a dictionary (i.e., type(dictionary) == dict)")
+    flat_dict = dict(_flatten_dict_gen(dictionary, parent_key, sep))
+    return flat_dict
+
+
+def get_dict_paths(dictionary: dict, path: list=[]):
```
Comment (yangkcatiu) [inline osometweet/wrangle.py:82]: I added a default value to `path` to make the tests run. But this is not the ideal solution, see https://docs.python-guide.org/writing/gotchas/

File: osometweet/wrangle.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    """
+    if not isinstance(path,list):
+        raise TypeError("`path` must be a list")
+
+    if not isinstance(dictionary, dict):
+        yield path
+    else:
+        yield from [new for key, val in dictionary.items() for new in get_dict_paths(val, path+[key])]
+
+def get_dict_val(dictionary:dict, key_list:list=[]):
```
Comment (yangkcatiu) [inline osometweet/wrangle.py:130]: Same here.

File: osometweet/wrangle.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        raise TypeError("`dictionary` must be of type `dict`")
+
+    if not isinstance(key_list, list):
+        raise TypeError("`key_list` must be of type `list`")
+
+    retval = dictionary
+    for k in key_list:
+
+        # If retval is not an iterable, we're going too deep
+        if not hasattr(retval, '__iter__'):
```
Comment (yangkcatiu) [inline osometweet/wrangle.py:201]: We might encounter the corner cases where retval is a string or a list, which will break the code.

I added a test case for this. The test case fails if you run it now.

**Review by yangkcatiu — CHANGES_REQUESTED** [review]

File: osometweet/wrangle.py (hunk 4 above)
Comment (mr-devs) [reply osometweet/wrangle.py:82]: Thanks!

Do you think that this is an issue for this function? It doesn't seem to display the "gotcha" that is described in the linked-to article but maybe I'm missing something?

File: osometweet/wrangle.py (hunk 6 above)
Comment (mr-devs) [reply osometweet/wrangle.py:201]: I made a change that I think deals with this - just made this line more specific. However, now I'm not sure if it breaks something else... 😅  This function always confuses me haha

**Comment by mr-devs** [comment]
Thanks for the great feedback @yangkcatiu. Check out line 218 from my latest commit and let me know if you think this solves the issue with the `get_dict_val` edge case that you highlighted. It now passes the tests but not sure if it breaks something else?

I also made some minor formatting changes after running pylint.

File: osometweet/wrangle.py (hunk 4 above)
Comment (yangkcatiu) [reply osometweet/wrangle.py]: Not sure if this is going to cause some problems. Let's keep it as it is for now.

File: osometweet/wrangle.py (hunk 6 above)
Comment (yangkcatiu) [reply osometweet/wrangle.py]: Should be good now.

**Review by yangkcatiu — APPROVED** [review]

**Comment by yangkcatiu** [comment]
Let's come back to this If we see new bugs in the future. I don't have any cases that would break the code for now.
